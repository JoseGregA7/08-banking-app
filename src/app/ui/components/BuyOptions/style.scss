.BuyOptions__wrapper {
    background-color: #f9f9f9;
    padding: 10px;
    border-radius: 8px;
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
    max-width: 400px;
    margin: 0 auto;
  }
  
  .BuyOptions__title {
    font-size: 1rem;
    font-weight: bold;
    text-align: center;
    margin-bottom: 8px;
    color: #333;
  }
  
  .BuyOptions__form {
    display: flex;
    flex-direction: column;
  }
  
  .BuyOptions__field {
    margin-bottom: 10px;
  }
  
  .BuyOptions__label {
    font-size: 0.8rem;
    font-weight: 600;
    margin-bottom: 5px;
    color: #444;
  }
  
  .BuyOptions__select,
  .BuyOptions__input {
    padding: 4px;
    font-size: 0.8rem;
    border: 1px solid #ddd;
    border-radius: 4px;
    width: 100%;
    box-sizing: border-box;
    transition: all 0.3s ease;
  }
  
  .BuyOptions__select:focus,
  .BuyOptions__input:focus {
    outline: none;
    border-color: #007BFF;
    box-shadow: 0 0 5px rgba(0, 123, 255, 0.2);
  }
  
  .BuyOptions__submitButton {
    padding: 8px;
    font-size: 0.8rem;
    background-color: #007BFF;
    color: white;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s ease;
  }
  
  .BuyOptions__submitButton:hover {
    background-color: #910910;
  }
  
  .BuyOptions__submitButton:disabled {
    background-color: #c0c0c0;
    cursor: not-allowed;
  }
  
  .BuyOptions__input[type="number"] {
    -moz-appearance: textfield;
    -webkit-appearance: none;
    appearance: none;
  }
  
  .BuyOptions__input[type="number"]::-webkit-outer-spin-button,
  .BuyOptions__input[type="number"]::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  