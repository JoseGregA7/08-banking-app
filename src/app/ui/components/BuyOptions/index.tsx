import './style.scss';

const BuyOptions = ({
  handleBuyTypeChange,
  handleBuyAmountChange,
  handleBuySubmit,
  state,
}: {
  handleBuyTypeChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  handleBuyAmountChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleBuySubmit: (e: React.FormEvent) => void;
  state: any;
}) => {
  return (
    <div className="BuyOptions__wrapper" data-testid="buy-options-wrapper">
      <h3 className="BuyOptions__title">Formulario de Compras</h3>
      <form
        onSubmit={handleBuySubmit}
        className="BuyOptions__form"
        aria-labelledby="buy-form-title"
      >
        <div className="BuyOptions__field">
          <label
            htmlFor="purchaseType"
            className="BuyOptions__label"
            id="buy-form-title"
          >
            Tipo de Compra:
          </label>
          <select
            id="purchaseType"
            value={state.buyType}
            onChange={handleBuyTypeChange}
            className="BuyOptions__select"
            aria-label="Seleccionar el tipo de compra"
            data-testid="buy-type-select"
          >
            <option value="PHYSICAL_ESTABLISHMENT">
              En Establecimiento Físico
            </option>
            <option value="WEB_PAGE">En la Página Web</option>
          </select>
        </div>
        <div className="BuyOptions__field">
          <label
            htmlFor="amount"
            className="BuyOptions__label"
            aria-label="Cantidad a comprar"
          >
            Cantidad a comprar:
          </label>
          <input
            id="amount"
            type="number"
            value={state.buyAmount}
            onChange={handleBuyAmountChange}
            min="0"
            className="BuyOptions__input"
            aria-labelledby="amount-label"
            aria-placeholder="Ingrese la cantidad de productos"
            data-testid="buy-amount-input"
          />
        </div>
        <button
          type="submit"
          className="BuyOptions__submitButton"
          aria-label="Realizar compra"
          data-testid="buy-submit-button"
        >
          Realizar Compra
        </button>
      </form>
    </div>
  );
};

export default BuyOptions;
