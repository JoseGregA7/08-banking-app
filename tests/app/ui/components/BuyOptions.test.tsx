import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';
import { describe, expect, test, vitest } from 'vitest';
import BuyOptions from '../../../../src/app/ui/components/BuyOptions';

describe('BuyOptions', () => {
  const mockHandleBuyTypeChange = vitest.fn();
  const mockHandleBuyAmountChange = vitest.fn();
  const mockHandleBuySubmit = vitest.fn();
  const state = {
    buyType: 'PHYSICAL_ESTABLISHMENT',
    buyAmount: 100,
  };

  test('should call handleBuySubmit when submit button is clicked', () => {
    const { getByTestId } = render(
      <BuyOptions
        handleBuyTypeChange={mockHandleBuyTypeChange}
        handleBuyAmountChange={mockHandleBuyAmountChange}
        handleBuySubmit={mockHandleBuySubmit}
        state={state}
      />
    );

    const formElement = getByTestId('buy-options-wrapper');
    expect(formElement).toBeInTheDocument(); 

    fireEvent.click(getByTestId('buy-submit-button'));

    expect(mockHandleBuySubmit).toHaveBeenCalledTimes(1);
  });

  test('should call handleBuyTypeChange when buy type is selected', () => {
    const { getByTestId } = render(
      <BuyOptions
        handleBuyTypeChange={mockHandleBuyTypeChange}
        handleBuyAmountChange={mockHandleBuyAmountChange}
        handleBuySubmit={mockHandleBuySubmit}
        state={state}
      />
    );

    fireEvent.change(getByTestId('buy-type-select'), {
      target: { value: 'WEB_PAGE' },
    });

    expect(mockHandleBuyTypeChange).toHaveBeenCalledTimes(1);
  });

  test('should call handleBuyAmountChange when amount is entered', () => {
    const { getByTestId } = render(
      <BuyOptions
        handleBuyTypeChange={mockHandleBuyTypeChange}
        handleBuyAmountChange={mockHandleBuyAmountChange}
        handleBuySubmit={mockHandleBuySubmit}
        state={state}
      />
    );

    fireEvent.change(getByTestId('buy-amount-input'), {
      target: { value: '200' },
    });

    expect(mockHandleBuyAmountChange).toHaveBeenCalledTimes(1);
  });

  test('should match snapshot', () => {
    const { asFragment } = render(
      <BuyOptions
        handleBuyTypeChange={mockHandleBuyTypeChange}
        handleBuyAmountChange={mockHandleBuyAmountChange}
        handleBuySubmit={mockHandleBuySubmit}
        state={state}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
