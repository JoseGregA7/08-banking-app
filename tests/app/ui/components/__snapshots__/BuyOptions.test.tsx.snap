// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BuyOptions > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="BuyOptions__wrapper"
    data-testid="buy-options-wrapper"
  >
    <h3
      class="BuyOptions__title"
    >
      Formulario de Compras
    </h3>
    <form
      aria-labelledby="buy-form-title"
      class="BuyOptions__form"
    >
      <div
        class="BuyOptions__field"
      >
        <label
          class="BuyOptions__label"
          for="purchaseType"
          id="buy-form-title"
        >
          Tipo de Compra:
        </label>
        <select
          aria-label="Seleccionar el tipo de compra"
          class="BuyOptions__select"
          data-testid="buy-type-select"
          id="purchaseType"
        >
          <option
            value="PHYSICAL_ESTABLISHMENT"
          >
            En Establecimiento Físico
          </option>
          <option
            value="WEB_PAGE"
          >
            En la Página Web
          </option>
        </select>
      </div>
      <div
        class="BuyOptions__field"
      >
        <label
          aria-label="Cantidad a comprar"
          class="BuyOptions__label"
          for="amount"
        >
          Cantidad a comprar:
        </label>
        <input
          aria-labelledby="amount-label"
          aria-placeholder="Ingrese la cantidad de productos"
          class="BuyOptions__input"
          data-testid="buy-amount-input"
          id="amount"
          min="0"
          type="number"
          value="100"
        />
      </div>
      <button
        aria-label="Realizar compra"
        class="BuyOptions__submitButton"
        data-testid="buy-submit-button"
        type="submit"
      >
        Realizar Compra
      </button>
    </form>
  </div>
</DocumentFragment>
`;
